### PR TITLE
Reward Boss implementation faithful to Tibia's one.

### PR DESCRIPTION
**Pull request by socket2810**
## Reasoning
After seeing a few reward systems on [OTLand](https://otland.net/) try and fail to replicate the original one, I decided to give it a try myself. Some images of the result:
Boss Loot Reward Chest       Reward Chest 2

### Boss Corpses
The difference between a normal monster corpse and a corpse that should yield individual rewards to players that participated in the fight is its attributes. **ITEM_ATTRIBUTE_CORPSEOWNER** is set to 2^31-1 and **ITEM_ATTRIBUTE_DATE** stores the time it was killed. 
### New class: Reward
It's the *Reward* class responsibility to store the different rewards a player may receive and, just like a depot chest, its contents are stored into the database when the player logs out and again loaded when they log back.
### New class: Reward Chest
Analogous to the Depot Locker, the *Reward Chest* contains the *Reward* (that inside the game is a [reward container](http://tibia.wikia.com/wiki/Reward_Container). However, it contains all the *Rewards*. They are inserted into the chest during the player load phase and every time another *Reward* is created via:
````
Player::getReward(uint32_t timestamp, bool autoCreate);
// or
Player.getReward(self, timestamp, autoCreate)
````

### Timestamps
Timestamps are set both on the boss corpse and on the in-game *reward container* / *Reward* object. They're used to identify the *Reward* and to limit how long the reward is stored in the database (Tibia chose a maximum of 7 days).
### How is each player attributed a different ~~corpse~~reward container?
~~The *Player* class now holds a *rewardCorpses* map that stores each corpse, which are instantiated the first time that the player opens the body and are collected in the player destructor (together with the player's *Rewards*).~~ (Not anymore: 7db6344)
The boss now drops a *proxy* reward container, meaning it is **not** a *Reward* object, but simply an item with the same id. Whenever a player tries to open this container, he will actually open his own *Reward*, this way there is no need to store unique corpses objects for each player.
### How is the loot handled?
Each player's contribution is measured by an *onHealthChange* event. The player will be considered **active** in the fight if he is currently in the boss' **targetList** - or whenever he heals someone that is currently active. The contribution is kept stored even if the player disconnects during the fight. If during the loot distribution phase the player is offline, he'll still receive his reward (Lua implementation: [insertRewardBag(guid, timestamp, itemList)](https://github.com/socket2810/forgottenserver/blob/master/data/creaturescripts/scripts/boss.lua#L188)).
### How to enable reward distribution to a boss
Set the **rewardboss* flag to true on the monster .xml file:
````xml
<flag rewardboss="1" />
````
To make a loot unique and awarded to the highest scoring player:
````xml
<item id="5903" chance="100000" unique="1"/><!-- ferumbras' hat -->
````
And register the corresponding events by calling *Monster.setReward(self, enable)* whenever a boss is summoned (e.g [place_monster.lua](https://github.com/socket2810/forgottenserver/blob/master/data/talkactions/scripts/place_monster.lua#L14)).
### Player contribution
Player contribution is currently calculated like this:
````lua
local score = (playerDamageOut / totalDamageOut) + (playerDamageIn / totalDamageIn) + (player.Healing / totalHealing)
score = score / 3 -- normalize to 0-1
local expectedScore = 1 / participants
local lootFactor = 1
lootFactor = lootFactor / participants ^ (1 / 3) -- tone down the loot a notch if there are many participants
lootFactor = lootFactor * (1 + lootFactor) ^ (score / expectedScore) -- increase the loot multiplicatively by how many times the player surpassed the expected score
````
Each *lootBlock* is therefore **lootFactor** times more probable to be awarded to a specific player. However it is far from being optimal if we want to achieve a result similar to Tibia's and *only* with the community help we will manage that.

### CODE REVIEW

**Comment by thexamx** [comment]
what about rares? And rest of loot? Because one loot should be sliced to all contributors in killing boss.

File: data/creaturescripts/creaturescripts.xml [hunk 1]
```diff
@@ -8,4 +8,8 @@
 	<event type="death" name="PlayerDeath" script="playerdeath.lua" />
 	<event type="death" name="DropLoot" script="droploot.lua" />
 	<event type="extendedopcode" name="ExtendedOpcode" script="extendedopcode.lua" />
+
+    <event type="death" name="BossDeath" script="boss.lua" />
+    <event type="healthchange" name="BossParticipation" script="boss.lua" />
+    <event type="think" name="BossThink" script="boss.lua" />
```
Comment (WibbenZ) [inline data/creaturescripts/creaturescripts.xml]: Indendation on all the lines

File: data/creaturescripts/scripts/logout.lua [hunk 2]
```diff
@@ -3,5 +3,18 @@ function onLogout(player)
 	if nextUseStaminaTime[playerId] ~= nil then
 		nextUseStaminaTime[playerId] = nil
 	end
+
+    local stats = player:inBossFight()
+    if(stats) then
```
Comment (WibbenZ) [inline data/creaturescripts/scripts/logout.lua]: Remove the parentheses aswell as line 11, 13
Aswell as indentation

File: data/events/scripts/player.lua [hunk 3]
```diff
@@ -180,3 +176,29 @@ function Player:onGainSkillTries(skill, tries)
...
+		for _, item in ipairs(tile:getItems()) do
+			if item:getAttribute(ITEM_ATTRIBUTE_CORPSEOWNER) == 2^31 - 1 and item:getName() == container:getName() then
+				self:sendCancelMessage('Sorry, not possible.')
+				return false
+			end
+		end
+	end
+
+	return true
+end
```
Comment (WibbenZ) [inline data/events/scripts/player.lua]: Missing EOF line

File: data/events/scripts/player.lua [hunk 4]
```diff
@@ -180,3 +176,29 @@ function Player:onGainSkillTries(skill, tries)
...
+		if(not container) then
+			return true 
+		end
+		local itemId = container:getId()
+		-- Do not let the player insert items into either the Reward Container or the Reward Chest
+		if itemId == 21518 or itemId == 21584 then
+			self:sendCancelMessage('Sorry, not possible.')
+			return false
+		end
+		-- The player also shouldn't be able to insert items into the boss corpse
```
Comment (WibbenZ) [inline data/events/scripts/player.lua]: New line

File: data/events/scripts/player.lua [hunk 5]
```diff
@@ -180,3 +176,29 @@ function Player:onGainSkillTries(skill, tries)
...
 end
+
+function Player:onMoveItem(item, count, fromPosition, toPosition)
+	if toPosition.x == CONTAINER_POSITION then
+		local cid = toPosition.y - 64 -- container id
+		local container = self:getContainerById(cid)		
+		if(not container) then
+			return true 
+		end
+		local itemId = container:getId()
```
Comment (WibbenZ) [inline data/events/scripts/player.lua]: New line

File: data/talkactions/scripts/place_monster.lua [hunk 6]
```diff
@@ -10,6 +10,9 @@ function onSay(player, words, param)
 	local position = player:getPosition()
 	local monster = Game.createMonster(param, position)
 	if monster ~= nil then
+	    if monster:getType():isRewardBoss() then
```
Comment (WibbenZ) [inline data/talkactions/scripts/place_monster.lua]: Indentation

File: src/luascript.cpp [hunk 7]
```diff
@@ -7605,6 +7613,62 @@ int LuaScriptInterface::luaPlayerGetFreeCapacity(lua_State* L)
...
+		return 1;
+	}
+
+	uint32_t rewardId = getNumber<uint32_t>(L, 2);
+	bool autoCreate = getBoolean(L, 3, false);	
+	if (Reward* reward = player->getReward(rewardId, autoCreate)) {
+		pushUserdata<Item>(L, reward);
+		setItemMetatable(L, -1, reward);
+	}
+	else {
```
Comment (WibbenZ) [inline src/luascript.cpp]: Indentation

File: src/luascript.cpp [hunk 8]
```diff
@@ -11391,6 +11455,19 @@ int LuaScriptInterface::luaMonsterTypeIsHealthShown(lua_State* L)
...
 }
 
+int LuaScriptInterface::luaMonsterTypeIsRewardBoss(lua_State* L)
+{
+	// monsterType:isRewardBoss()
+	MonsterType* monsterType = getUserdata<MonsterType>(L, 1);
+	if (monsterType) {
+		pushBoolean(L, monsterType->isRewardBoss);
+	}
+	else {
```
Comment (WibbenZ) [inline src/luascript.cpp]: Indentation

File: src/reward.cpp [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+	}
+}
+
+Cylinder* Reward::getParent() const
+{
+	if (parent) {
+		return parent->getParent();
+	}
+	return nullptr;
+}
```
Comment (WibbenZ) [inline src/reward.cpp]: Missing EOF line

File: src/reward.h [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+	}
+
+	Cylinder* getParent() const final;
+	Cylinder* getRealParent() const final {
+		return parent;
+	}
+};
+
+#endif
+
```
Comment (WibbenZ) [inline src/reward.h]: Remove the extra EOF line

File: src/rewardchest.cpp [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+		parent->postAddNotification(thing, oldParent, index, LINK_PARENT);
+	}
+}
+
+void RewardChest::postRemoveNotification(Thing* thing, const Cylinder* newParent, int32_t index, cylinderlink_t)
+{
+	if (parent != nullptr) {
+		parent->postRemoveNotification(thing, newParent, index, LINK_PARENT);
+	}
+}
```
Comment (WibbenZ) [inline src/rewardchest.cpp]: Missing EOF line

File: src/rewardchest.h [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+	void postAddNotification(Thing* thing, const Cylinder* oldParent, int32_t index, cylinderlink_t link = LINK_OWNER) final;
+	void postRemoveNotification(Thing* thing, const Cylinder* newParent, int32_t index, cylinderlink_t link = LINK_OWNER) final;
+
+	bool canRemove() const final {
+		return false;
+	}
+};
+
+#endif
+
```
Comment (WibbenZ) [inline src/rewardchest.h]: Remove the extra EOF line

File: data/creaturescripts/scripts/boss.lua [hunk 13]
```diff
@@ -0,0 +1,376 @@
+local attr = {
```
Comment (WibbenZ) [inline data/creaturescripts/scripts/boss.lua]: Indendation on pretty much all lines?

File: data/creaturescripts/scripts/boss.lua [hunk 14]
```diff
@@ -0,0 +1,376 @@
...
+        local healerStats = getPlayerStats(stats.bossId, attacker:getGuid(), true)
+        healerStats.active = true
+        healerStats.playerId = attackerId -- Update player id
+        healerStats.healing = healerStats.healing + primaryDamage
+    elseif stats.bossId == attackerId then
+        -- Account for damage taken from the boss
+        stats.damageIn = stats.damageIn + primaryDamage
+    end
+    return primaryDamage, primaryType, secondaryDamage, secondaryType
+end
```
Comment (WibbenZ) [inline data/creaturescripts/scripts/boss.lua]: Missing EOF line

**Comment by Healek** [comment]
Error in compile, how to fix?

/theforgottenserver/src/items.h:59:2: error: expected '}' before 'ITEM_TYPE_REWARDCHEST'
  ITEM_TYPE_REWARDCHEST,
  ^
/theforgottenserver/src/items.h:60:2: error: 'ItemTypes_t ITEM_TYPE_LAST' redeclared as different kind of symbol
  ITEM_TYPE_LAST,
  ^
/theforgottenserver/src/items.h:58:2: error: previous declaration of 'ItemTypes_t ITEM_TYPE_LAST'
  ITEM_TYPE_LAST
  ^
/theforgottenserver/src/items.h:61:1: error: expected unqualified-id before '}' token
 };
 ^
/castmaster/src/items.h:61:1: error: expected declaration before '}' token
make[2]: *** [CMakeFiles/tfs.dir/src/actions.cpp.o] Error 1
make[1]: *** [CMakeFiles/tfs.dir/all] Error 2
make: *** [all] Error 2

**Comment by Healek** [comment]
Error in compile, how to fix?

/theforgottenserver/src/items.h:59:2: error: expected '}' before 'ITEM_TYPE_REWARDCHEST'
ITEM_TYPE_REWARDCHEST,
^
/theforgottenserver/src/items.h:60:2: error: 'ItemTypes_t ITEM_TYPE_LAST' redeclared as different kind of symbol
ITEM_TYPE_LAST,
^
/theforgottenserver/src/items.h:58:2: error: previous declaration of 'ItemTypes_t ITEM_TYPE_LAST'
ITEM_TYPE_LAST
^
/theforgottenserver/src/items.h:61:1: error: expected unqualified-id before '}' token
};
^
/theforgottenserver/src/items.h:61:1: error: expected declaration before '}' token
make[2]: *** [CMakeFiles/tfs.dir/src/actions.cpp.o] Error 1
make[1]: *** [CMakeFiles/tfs.dir/all] Error 2
make: *** [all] Error 2

**Comment by Healek** [comment]
error in compilation, how to fix?

/theforgottenserver/src/actions.cpp: In member function 'ReturnValue Actions::internalUseItem(Player*, const Position&, uint8_t, Item*, bool)':
/theforgottenserver/src/actions.cpp:348:18: error: 'class Container' has no member named 'isRewardCorpse'
   if (container->isRewardCorpse()) {
                  ^
make[2]: *** [CMakeFiles/tfs.dir/src/actions.cpp.o] Error 1
make[1]: *** [CMakeFiles/tfs.dir/all] Error 2
make: *** [all] Error 2

**Comment by conde2** [comment]
@socket2810 

Just curious if there is any reason for closing this PR...

**Comment by socket2810** [comment]
@conde2 

I had made it in my master branch. Whenever I make any new changes I'll open another PR and reference this one.

**Comment by ranisalt** [comment]
You could as well rename your branch locally, push it to github, then reset against this master...

**Comment by socket2810** [comment]
@ranisalt
I hope you can forgive me if this was the only solution I could find with my limited time and git knowledge...

**Pull request closed without merging**